chalk-solve/src/coherence.rs

```diff
@@ -41,19 +41,25 @@ impl std::error::Error for CoherenceError {}
 /// Stores the specialization priorities for a set of impls.
 /// This basically encodes which impls specialize one another.
 #[derive(Clone, Debug, Default, PartialEq, Eq)]
-pub struct SpecializationPriorities {
-    map: BTreeMap<ImplId, SpecializationPriority>,
+pub struct SpecializationPriorities<TF: TypeFamily> {
+    map: BTreeMap<ImplId<TF>, SpecializationPriority>,
 }
 
-impl SpecializationPriorities {
+impl<TF: TypeFamily> SpecializationPriorities<TF> {
+    pub fn new() -> Self {
```

Comment:
If `SpecializationPriorities` derives `Default`, why is this needed?

Reply:
True, we don't need this, we can write `::default`...

Reply:
Ah, I remember. The problem is that `TF` doesn't necessarily implement `Default`. The fix would be either:

* to manually implement `Default` for `SpecializationPriorities<TF>` (for any `TF`, not just `TF: Default`)
* to require that all `TypeFamily: Default`, as we did with `Clone` and various other traits

The former is more typing that `new` which I think is why I didn't do it. The latter is actually a fine strategy so long as type families remain "dummy" types that just pull together a bunch of other logic. But I think that eventually we will want a `TypeFamily` to actually *be* an arena, in which case it may not want to implement `Default`. (i.e., in rustc, I expect a type family to be the `tcx`). That said, we could also make it so that you have `TF::Context` to represent the arena value -- but I think that may be more annoying in the end. I thought it'd be useful if we actually make a `TF` be the context itself, since that will help with type inference.

(I contemplated renaming `TF` to `TCX`... but I'm not sure we plan to keep that name in the compiler.)